pkg/daemon/ceph/client/pool.go

```diff
@@ -418,6 +421,68 @@ func CreateReplicatedPoolForApp(context *clusterd.Context, clusterInfo *ClusterI
 	}
 
 	logger.Infof("creating replicated pool %s succeeded", poolName)
+
+	if checkFailureDomain {
+		if err = ensureFailureDomain(context, clusterInfo, clusterSpec, poolName, pool); err != nil {
+			return nil
+		}
+	}
+	return nil
+}
+
+func ensureFailureDomain(context *clusterd.Context, clusterInfo *ClusterInfo, clusterSpec *cephv1.ClusterSpec, poolName string, pool cephv1.PoolSpec) error {
+	if pool.FailureDomain == "" {
+		logger.Debugf("skipping check for failure domain on pool %q as it is not specified", poolName)
+		return nil
+	}
+
+	logger.Debugf("checking that pool %q has the failure domain %q", poolName, pool.FailureDomain)
+	details, err := GetPoolDetails(context, clusterInfo, poolName)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get pool %q details", poolName)
+	}
+
+	// Find the failure domain for the current crush rule
+	rule, err := getCrushRule(context, clusterInfo, details.CrushRule)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get crush rule %q", details.CrushRule)
+	}
+	currentFailureDomain := ""
+	if len(rule.Steps) < 2 || rule.Steps[1].Type == "" {
+		logger.Warningf("unexpected format for crush rule %q, proceeding to create a new crush rule", details.CrushRule)
+	} else if rule.Steps[1].Type != pool.FailureDomain {
+		currentFailureDomain = rule.Steps[1].Type
+	} else {
+		logger.Debugf("pool %q has the expected failure domain %q", poolName, pool.FailureDomain)
+		return nil
+	}
+
+	// Use a crush rule name that is unique to the desired failure domain
+	crushRuleName := fmt.Sprintf("%s_%s", poolName, pool.FailureDomain)
+	logger.Infof("updating pool %q failure domain from %q to %q with new crush rule %q", poolName, currentFailureDomain, pool.FailureDomain, crushRuleName)
+	logger.Infof("crush rule %q will no longer be used by pool %q", details.CrushRule, poolName)
+
+	// Create a new crush rule for the expected failure domain
+	if err := createReplicationCrushRule(context, clusterInfo, clusterSpec, crushRuleName, pool); err != nil {
```

Comment:
What if someone reverts to the previous failure domain? Is the crush create command idempotent? Or does it return `EEXIST`? Should we check for the rule existence first?

Reply:
It's idempotent, no error is returned, so instead of checking for existence it seems sufficient to run again.